### ISSUE TITLE
Remote computer goes to sleep during win_updates being applied

### ISSUE THREAD

**Issue by agibson2** [ansible-collections/ansible.windows#310]
##### SUMMARY
I am not sure if this is considered a bug or a feature request.  It appears that the win_updates process doesn't make sure the computer doesn't go to sleep during the updates.

This is with the latest version using pip and ansible-galaxy on CentOS 7 with RPM packages removed of ansible and using pip and ansible-galaxy to install ansible-core and ansible.windows. While doing updates, sometime in the middle or near the end, the computer went to sleep causing the updates to pause and giving an error for win_updates because the computer was asleep.

Error is below...

```TASK [Apply updates] ***************************************************************************************************
[WARNING]: Unknown failure when polling update result - attempting to cancel task: ('Connection aborted.', error(111,
'Connection refused'))
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ConnectionError: ('Connection aborted.', error(111, 'Connection refused'))
fatal: [test-pc2]: FAILED! => {"msg": "Unexpected failure during module execution.", "stdout": ""}
```

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
win_updates

##### ANSIBLE VERSION

```
(pipansible) [root@Server-PC Ansible]# ansible --version
[DEPRECATION WARNING]: Ansible will require Python 3.8 or newer on the controller starting with Ansible 2.12. Current
version: 2.7.5 (default, Nov 16 2020, 22:23:17) [GCC 4.8.5 20150623 (Red Hat 4.8.5-44)]. This feature will be removed
from ansible-core in version 2.12. Deprecation warnings can be disabled by setting deprecation_warnings=False in
ansible.cfg.
/root/pipansible/lib/python2.7/site-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
ansible [core 2.11.6]
  config file = /root/.ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /root/pipansible/lib/python2.7/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /root/pipansible/bin/ansible
  python version = 2.7.5 (default, Nov 16 2020, 22:23:17) [GCC 4.8.5 20150623 (Red Hat 4.8.5-44)]
  jinja version = 2.11.3
  libyaml = True
```

##### COLLECTION VERSION

```
Collection      Version
--------------- -------
ansible.windows 1.7.3
```

##### CONFIGURATION

```
DEFAULT_LOG_PATH(/root/.ansible.cfg) = /root/ansible.log
```

##### OS / ENVIRONMENT
CentOS 7  (rpm packages of ansible removed and ansible-core installed vi pip in virtualenv as well as using ansible-galaxy to install ansible.windows collection)

##### STEPS TO REPRODUCE
Set computer to go to sleep in 2 minutes on a Windows 10 desktop.  Start a win_updates to apply windows updates.  Computer goes to sleep before computer is finished installed and rebooted.  Error appears on the controller computer because computer is asleep.
```

```

##### EXPECTED RESULTS
I would have expected the computer sleep to be suppressed during the update to make sure it finishes.  Maybe that is expecting too much though.  I am not sure if this is considered a bug or a feature request.  I am assuming there is a way to tell the PC to not sleep until the computer is rebooted. 

##### ACTUAL RESULTS

```paste below

```

**Comment by jborean93** [comment]
Thanks for the bug report. I'm not sure if this is going to be possible to do without editing the system power settings which may or may not be desirable. I'll have to look further into this to see what's possible and figure out if there is a way around this.

**Comment by jborean93** [comment]
A quick search brings up [SetThreadExecutionState](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setthreadexecutionstate) which can be called with `ES_CONTINUOUS | ES_SYSTEM_REQUIRED`. This seems like it will be workable for this scenario so will look at implementing that.

**Comment by agibson2** [comment]
Sounds great.  Thanks for looking at it.  Servers of course likely wouldn't have that issue but desktops are much more likely to sleep.  A wake packet can be sent to them to wake them up but they need to stay awake during the updates.

**Comment by jborean93** [comment]
If you get a chance are you able to try out https://github.com/ansible-collections/ansible.windows/pull/311 and see if it works on your system?

**Comment by agibson2** [comment]
I posted this previously to the merge request I think instead of this issue...

I replaced the win_updates.ps1, uninstalled the last windows cumulative update for the win10 system, rebooted. I then set the sleep time to 3 minutes. Now the win10 system is in a constant reboot cycle trying to apply updates using ansible. Ansible is still applying updates over and over it seems. I logged in to make sure there was activity to make sure the reboot cycle wasn't related to a too quick sleep timeout but it still keeps rebooting. Likely not related to this change I assume. I didn't see it sleep yet but the reboot happens within 2 or 3 minutes so I don't think it gets to that part.

This is the win_updates log from the win10 system....
https://pastebin.com/bGswAEx8

additionally...  I canceled the ansible win_updates and then logged into the computer.  It only showed a windows defender update.  I applied that and then clicked check for updates and it found the Cumulative Update again and installed it automatically.  Strange behavior for windows updates and Ansible.

After it applied the updates, I will remove the update again and then wait several hours (maybe until tomorrow).  by that time, It likely will have rechecked for updates and not get into that state again.  I can then do the sleep testing hopefully.

**Comment by agibson2** [comment]
I have noticed that before where applying updates using Ansible doesn't update whatever the GUI has cached and thinks need to be done so they are out of sync for awhile until Windows decides to check again on it's own.  I have seen where Ansible has updated everything but the GUI still thinks there are updates to apply after Ansible reboots and finishes.  Not sure what OSs exhibit that behavior or if it repeatable every time.

**Comment by agibson2** [comment]
I did a system restore to before windows updates was applied.  I did a --check and it saw the updates that were needed.  I ran the update with the new ps1 code and it did keep the system from sleeping so that part is working.  It is now in the same reboot loop though.  I will back go back to the original win_updates.ps1 that was put in by ansible-galaxy and see if the same behavior is there after I do another restore to before the updates.

**Comment by agibson2** [comment]
Well the reboot loop is still there with the original ansible.windows 1.7.3 win_updates.ps1 version so that is a separate issue.  The updates did apply without the computer going to sleep so the added code does appear to fix that.  Not sure what is up with the constant reboots though...  Ill keep experimenting and see if it eventually stops if I give it a break for a day or so.  That is outside the scope of this issue though I think.

**Comment by agibson2** [comment]
I waited several hours, updated to ansible.windows 1.8.0 and updates applied without a the reboot loop.  I think the waiting is the key that made it work and not updating to 1.8.0.  I verified that the computer goes to sleep with that version still.

Next testing will be to do a system restore before windows updates this month again, Try to update the system within 5 or 10 minutes when all disk activity goes away after the restore, and see if the reboot loop is still there with ansible.windows 1.8.0.  If it is, then I will wait 6 hours or so and try it again.  That will reveal if 1.8.0 fixes the reboot issue or if just waiting several hours fixes it.

The behavior when the reboot issue is there (separate issue I think) shows that it adds the update, tries installing it but within 7 seconds gets an install progress update resultcode 1 with PercentComplete 100, CurrentUpdatePercentComplete 100 and rebootrequired false.  Immediately after that (same second) another received install progress update with ResultCode 2, percent complete 100, currentupdatepercentcomplete 100, and rebootrequired *true*.  It reboots the system and just keeps looping with the same 2 install progress updates after every reboot until I cancel the update.

**Comment by agibson2** [comment]
Reboot loop is still there with the windows.ansible 1.8.0 code after the restore (just like 1.7.4).  So it appears to fix the reboot loop... you just have to wait several hours.  I will wait until tomorrow afternoon, change the code to use your new .ps1 code, set the computer to sleep after 3 minutes and then start the ansible win_updates.  It should actually install the Cumulative update at that time and not go to sleep.

The reboot loop issue is odd.  Maybe there is some status that is coming back that could indicate what is going on that is not being looked at currently in the code?  That again seems to be a separate issue.  All indications with my testing is that will resolve itself several hours on it's own when windows refreshes something... likely when it checks automatically to see if any new udpates are available (automatic install is disabled... it only checks).

**Comment by agibson2** [comment]
@jborean93 I have verified again that the change you made does allow the updates to complete and win_updates to finish with a short computer sleep time (3 minutes).  I also verified that it fails with the 1.8.0 ansible.windows version that doesn't have your newest change.

````
TASK [Apply updates] *****************************************************************************************
changed: [test-win10]

TASK [Show updates] ******************************************************************************************
ok: [test-win10] => {
    "msg": [
        "Installed: 2021-11 Cumulative Update for Windows 10 Version 21H1 for x64-based Systems (KB5007186)",
        "Filtered:Microsoft Silverlight (KB4481252) (skip_optional)",
        "Filtered:Hewlett-Packard  - Other hardware, Printer - Null Print - HP Officejet Pro 8620 (category_names)"
    ]
}

PLAY RECAP ***************************************************************************************************
test-win10                   : ok=4    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
````


The constant reboot thing... I tried about 6 hours after and it was still doing the reboot loop showing success install in the logs but it was not.  About 2 hours later, I cleared the %windir%\SoftwareDistribution\DataStore directory (stopping windows update service before and restarting it after clearing the directory) on the Win10 system and it installed the update properly with win_updates on the next test I did.  I am not 100% sure if just waiting the additional 2 hours is what made it work or clearing that directory.  It did install the cumulative update the next time I tried though.

**Comment by jborean93** [comment]
Sorry for the delay in replying. First off thanks for testing the PR and confirming if fixes your issue. I'll merge that momentarily.

As for your 2nd problem at a guess it sounds like Windows is rolling back the update after the reboot due to some unknown failure. This process is all opaque to Ansible as it's just waiting for the reboot to finish and cannot see what the reboot is actually waiting on. The only way to confirm this is to look at the console of the host itself to see what is actually happening. It could also just be some problem with the update cache where Windows never sees it as installed and will keep on doing it until either certain time has passed or your clear the cache.

In the end I'm not sure if `win_updates` can actually fix this problem because it's at the mercy of what Windows reports through the known APIs. What we can do instead is potentially track the updates installed pre reboot and then the updates installed post reboot to determine if it's going to be stuck in an inifinite loop. It's not an ideal solution but at least it stops Ansible from this looping issue.

I recommend you open a separate issue to track this problem separate from this one which is about the sleep issue.
